tests/unit/test-wcs-notifications.php

```diff
@@ -3,180 +3,2184 @@
+	protected static function wcs_update_dates_after_early_renewal( $subscription, $early_renewal ) {
+		$dates_to_update = self::get_dates_to_update( $subscription );
 
+		if ( ! empty( $dates_to_update ) ) {
+			// translators: %s: order ID.
+			$order_number = sprintf( _x( '#%s', 'hash before order number', 'woocommerce-subscriptions' ), $early_renewal->get_order_number() );
+			$order_link   = sprintf( '<a href="%s">%s</a>', esc_url( wcs_get_edit_post_link( $early_renewal->get_id() ) ), $order_number );
+
+			try {
+				$subscription->update_dates( $dates_to_update );
+
+				// translators: placeholder contains a link to the order's edit screen.
+				$subscription->add_order_note( sprintf( __( 'Customer successfully renewed early with order %s.', 'woocommerce-subscriptions' ), $order_link ) );
+			} catch ( Exception $e ) {
+				// translators: placeholder contains a link to the order's edit screen.
+				$subscription->add_order_note( sprintf( __( 'Failed to update subscription dates after customer renewed early with order %s.', 'woocommerce-subscriptions' ), $order_link ) );
+			}
+		}
 	}
 
 	/**
-	 * Check that developers can customize email notification.
+	 * This fn is in subscriptions as \WCS_Early_Renewal_Manager::get_dates_to_update, but not in subscriptions core.
 	 *
-	 * @return void
+	 * @param $subscription
+	 *
+	 * @return array
 	 */
-	public function test_customize_email_content() {
+	public static function get_dates_to_update( $subscription ) {
+		$next_payment_time = $subscription->get_time( 'next_payment' );
+		$dates_to_update   = array();
+
+		if ( $next_payment_time > 0 && $next_payment_time > time() ) {
+			$next_payment_timestamp = wcs_add_time( $subscription->get_billing_interval(), $subscription->get_billing_period(), $next_payment_time );
 
+			if ( $subscription->get_time( 'end' ) === 0 || $next_payment_timestamp < $subscription->get_time( 'end' ) ) {
+				$dates_to_update['next_payment'] = gmdate( 'Y-m-d H:i:s', $next_payment_timestamp );
+			} else {
+				// Delete the next payment date if the calculated next payment date occurs after the end date.
+				$dates_to_update['next_payment'] = 0;
+			}
+		} elseif ( $subscription->get_time( 'end' ) > 0 ) {
+			$dates_to_update['end'] = gmdate( 'Y-m-d H:i:s', wcs_add_time( $subscription->get_billing_interval(), $subscription->get_billing_period(), $subscription->get_time( 'end' ) ) );
+		}
+
+		return $dates_to_update;
 	}
 
 	/**
-	 * Check that enabling and disabling of notifications works.
+	 * Wrapper around \WC_Subscription::delete_date that also returns the subscription (which we need in the tests).
 	 *
-	 * @return void
+	 * @param $subscription
+	 * @param $date_type
+	 *
+	 * @return WC_Subscription
 	 */
-	public function test_enable_disable_notifications() {
+	protected static function delete_subscription_date( $subscription, $date_type ) {
+		$subscription->delete_date( $date_type );
+		$subscription->save(); // Is this needed?
```

Comment:
I think not. Most probably removing that will resolve the issue in the `delete_date()` function of the scheduler controller.

Reply:
Seeing that e.g. `update_status` calls a bunch of `update_dates` and `delete_date` and after that also `WC_Subscription::save`, it might be needed actually?

Reply:
Sure, our tests should still work, so I removed it.